src/VirtualContent.js

```diff
@@ -0,0 +1,341 @@
+  }
+
+  [_intersectionObserverCallback](entries) {
+    for (const entry of entries) {
+      const target = entry.target;
+      const isIntersecting = entry.isIntersecting;
+
+      // Update if this element has moved into or out of the viewport.
+      if (target === this) {
+        this[_scheduleUpdate]();
+        break;
+      }
+
+      const targetParent = target.parentNode;
+
+      // Update if an empty space sentinel has moved into the viewport.
+      if (targetParent === this[_emptySpaceSentinelContainer] && isIntersecting) {
+        this[_scheduleUpdate]();
+        break;
+      }
+
+      // Update if a child has moved out of the viewport.
+      if (targetParent === this && !isIntersecting) {
+        this[_scheduleUpdate]();
+        break;
+      }
+    }
+  }
+
+  [_mutationObserverCallback](records) {
+    const estimatedHeights = this[_estimatedHeights];
+
+    for (const record of records) {
+      for (const node of record.removedNodes) {
+        if (node.nodeType === Node.ELEMENT_NODE) {
+          // Removed children should have be made visible again and we should
+          // unobserve them with the resize observer.
+          this[_resizeObserver].unobserve(node);
+          this[_intersectionObserver].unobserve(node);
+          node.removeAttribute('invisible');
+          estimatedHeights.delete(node);
+        }
+      }
+
+      for (const node of record.addedNodes) {
+        if (node.nodeType === Node.ELEMENT_NODE) {
+          // Added children should be invisible initially.
+          node.setAttribute('invisible', '');
+          estimatedHeights.set(node, DEFAULT_HEIGHT_ESTIMATE);
+        } else {
+          // Remove non-element children because we can't control their
+          // invisibility state or even prevent them from being rendered using
+          // CSS (they aren't distinctly selectable).
+
+          // These records are not coalesced, so test that the node is actually
+          // a child of this node before removing it.
+          if (node.parentNode === this) {
+            this.removeChild(node);
+          }
+        }
+      }
+    }
+
+    this[_scheduleUpdate]();
+  }
+
+  [_resizeObserverCallback]() {
+    this[_scheduleUpdate]();
+  }
+
+  [_onActivateinvisible](e) {
+    // Find the child containing the target and synchronously update, forcing
+    // that child to be visible. The browser will automatically scroll to that
+    // element because it is visible, which will trigger another update to make
+    // the surrounding nodes visible.
+    const child = e.target;
+    while (child.parentNode !== this) {
+      child = child.parentNode;
+    }
+    this[_update]({forceVisible: new Set([child])});
+  }
+
+  [_scheduleUpdate]() {
+    if (this[_updateRAFToken] !== undefined) return;
+
+    this[_updateRAFToken] = window.requestAnimationFrame(this[_update]);
+  }
+
+  [_update]({forceVisible = new Set()} = {}) {
```

Comment:
This method is pretty enormous. Is it possible to split it up into other methods? https://refactoring.guru/smells/long-method

For example, the closures you create could be methods that take extra arguments.

In general I wouldn't worry to much about symbol-naming everything; private methods should ship in V8 soon-ish, so underscore-prefixed is good enough for now.

Reply:
Going to leave as a to-do.